subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/GetOutputFilesVisitor.java

```diff
@@ -16,45 +16,63 @@
 
 package org.gradle.api.internal.tasks.properties;
 
+import com.google.common.base.Function;
 import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Iterables;
 import com.google.common.collect.Iterators;
 import org.gradle.api.NonNullApi;
 import org.gradle.api.internal.tasks.CacheableTaskOutputFilePropertySpec;
 import org.gradle.api.internal.tasks.CompositeTaskOutputPropertySpec;
 import org.gradle.api.internal.tasks.TaskOutputFilePropertySpec;
+import org.gradle.api.internal.tasks.TaskPropertySpec;
 import org.gradle.api.internal.tasks.TaskPropertyUtils;
+import org.gradle.internal.Factory;
 
+import javax.annotation.Nullable;
 import java.io.File;
 import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
 
 @NonNullApi
 public class GetOutputFilesVisitor extends PropertyVisitor.Adapter {
-    private List<TaskOutputFilePropertySpec> specs = new ArrayList<TaskOutputFilePropertySpec>();
+    private List<TaskOutputFilePropertySpec> declaredOutputFileProperties = new ArrayList<TaskOutputFilePropertySpec>();
     private ImmutableSortedSet<TaskOutputFilePropertySpec> fileProperties;
     private boolean hasDeclaredOutputs;
 
     @Override
     public void visitOutputFileProperty(TaskOutputFilePropertySpec outputFileProperty) {
         hasDeclaredOutputs = true;
-        if (outputFileProperty instanceof CompositeTaskOutputPropertySpec) {
-            Iterators.addAll(specs, ((CompositeTaskOutputPropertySpec) outputFileProperty).resolveToOutputProperties());
-        } else {
-            if (outputFileProperty instanceof CacheableTaskOutputFilePropertySpec) {
-                File outputFile = ((CacheableTaskOutputFilePropertySpec) outputFileProperty).getOutputFile();
-                if (outputFile == null) {
-                    return;
-                }
-            }
-            specs.add(outputFileProperty);
-        }
+        declaredOutputFileProperties.add(outputFileProperty);
     }
 
-    public ImmutableSortedSet<TaskOutputFilePropertySpec> getFileProperties() {
-        if (fileProperties == null) {
-            fileProperties = TaskPropertyUtils.collectFileProperties("output", specs.iterator());
-        }
-        return fileProperties;
+    public Factory<ImmutableSortedSet<TaskOutputFilePropertySpec>> getFilePropertiesFactory() {
```

Comment:
Why does this need to be lazily evaluated?

Reply:
If we don't evaluate this lazily, some exceptions happen during task graph building and not when actually selecting the task for execution. See these build results: https://builds.gradle.org/viewLog.html?buildId=9985576&tab=buildResultsDiv&buildTypeId=Gradle_Check_Stage_QuickFeedbackLinuxOnly_Trigger

Reply:
Is this just cosmetic change (different root exceptions, but the bottom-most exception is the same), or do we get different exceptions, maybe even get errors when we shouldn't?

Reply:
There are some exceptions which are cosmetic changes.

What concerns me the most is `DefaultTaskExecutionPlanParallelTest#handles an exception while while walking the task graph when an enforced task is present`:
 https://github.com/gradle/gradle/blob/4a62138f50192e3f1d5be5882d7fbcdbc3bc5b04/subprojects/core/src/test/groovy/org/gradle/execution/taskgraph/DefaultTaskExecutionPlanParallelTest.groovy#L730-L746
IIUC, the finalizer task should be skipped if there are some problems when determining its dependencies. I the outputs are evaluated eagerly, then the build fails with 
```
org.gradle.api.internal.tasks.TaskDependencyResolveException: Could not determine the dependencies of task ':finalizer'.
```